lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -136,15 +146,246 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
    */
   auto update_criteria = get_default_update_criteria();
   session_state.set_fsm_state(CREATED, update_criteria);
-  MLOG(MDEBUG) << "State of session changed to "
-               << session_fsm_state_to_str(session_state.get_state());
+  uint32_t cur_version = session_state.get_current_version();
+  session_state.set_current_version(++cur_version, update_criteria);
+  MLOG(MINFO) << "State of session changed to "
+               << session_fsm_state_to_str(session_state.get_state())
+	       << " Current version is " << session_state.get_current_version();
+
   /* Update the m5gsm_cause and prepare for respone along with actual cause*/
   prepare_response_to_access(
       imsi, session_state, magma::lte::M5GSMCause::OPERATION_SUCCESS);
-  pipelined_client_->set_upf_session(sess_info, call_back_void_upf);
+
+  SessionState::SessionInfo sess_info;
+  sess_info.imsi       = imsi;
+  sess_info.ip_addr    = ip_addr;
+  sess_info.Pdr_rules_ = session_state.get_all_pdr_rules();
+  sess_info.Far_rules_ = session_state.get_all_far_rules();
+  session_state.sess_infocopy(&sess_info);
+  pipelined_client_->set_upf_session(sess_info, call_back_upf);
   return;
 }
 
+/* Function to initiate release of the session in enforcer requested by AMF
+ * Go over session map and find the respective session of imsi
+ * Go over SessionState vector and find the respective dnn (apn) 
+ * start terminating session process
+ */
+bool SessionStateEnforcer::m5g_release_session(
+      SessionMap& session_map, const std::string& imsi,
+      const std::string& dnn, SessionUpdate& session_update)
+{
+  /* Search with session search criteria of IMSI and apn/dnn and 
+   * find  respective sesion to release operation
+   */
+  SessionSearchCriteria criteria(imsi, IMSI_AND_APN, dnn);
+  auto session_it = session_store_.find_session(session_map, criteria);
+  if(!session_it) {
+     MLOG(MERROR) << "No session found in SessionMap for IMSI " << imsi
+                  << " with DNN " << dnn << " to release"; 
+     return false;
+  } else {

```

Comment:
We don't seem to need this else clause if we are returning false eariler :)

Reply:
fixed, corrected.